### PR TITLE
Adding documentation on K3s testing

### PR DESCRIPTION
**Pull request by dereknola**
#### Proposed Changes ####
This adds documentation describing how tests should be created and run in K3s.
Created integration subfolder of tests, now how e2e and integration under one clear roof.
Consolidated testing utilities under the tests folder.
Fixed typos in README.md markdown

#### Types of Changes ####
Documentation and Testing

#### Verification ####
N/A

#### Linked Issues ####
https://github.com/k3s-io/k3s/issues/3729

#### User-Facing Change ####

```release-note
NONE
```

### CODE REVIEW

File: pkg/etcd/etcd_int_test.go [hunk 1]
```diff
@@ -9,76 +9,76 @@ import (
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	"github.com/rancher/k3s/pkg/util/tests"
+	"github.com/rancher/k3s/tests/testutil"
```
Comment (briandowns) [inline pkg/etcd/etcd_int_test.go:12]: Not a blocker for merging: This reminds me, this package name should be changed to remove the redundancy. I thought I had commented that but I guess I didn't. We can use an import alias to disambiguate if necessary.

File: tests/testutil/cmd.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package tests
+package testutil
```
Comment (briandowns) [inline tests/testutil/cmd.go:1]:
```suggestion
package util
```

File: tests/testutil/runtime.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package tests
+package testutil
```
Comment (briandowns) [inline tests/testutil/runtime.go:1]:
```suggestion
package util
```

File: pkg/etcd/etcd_int_test.go (hunk 1 above)
Comment (briandowns) [reply pkg/etcd/etcd_int_test.go:12]: Commented below at the name change.

File: tests/TESTING.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+Go testing in K3s comes in 3 forms: Unit, Integration, and End-to-End (E2E). This
+document will explain *when* each test should be written and *how* each test should be
+generated, formatted, and run.
+
+Note: all shell commands given are relateive to the root k3s repo directory.
+___
+
+## Unit Tests
+
+Unit tests should be written when a complex component or function of a package needs testing.
```
Comment (briandowns) [inline tests/TESTING.md:12]: I'm not sure we want to say "complex" here as there are going to be trivial and maybe more obvious changes made that are in the critical path that also need testing.

File: tests/TESTING.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+### Framework
+
+All unit tests in K3s follow a [Table Driven Test](https://github.com/golang/go/wiki/TableDrivenTests) style. Specifically, K3s unit tests are automatically generated using the [gotests](https://github.com/cweill/gotests) tool. This is built into the Golang vscode extension or can be run via command line. Additionally, a set of custom templates is provided to the tool to extend the generated test's functionality. To use these templates, call:
+
+```bash
+gotests --template_dir=<PATH_TO_K3S>/contrib/gotests_templates
+```
+
+Or in vscode, edit the golang extension setting `Go: Generate Tests Flags`  
+and add `--template_dir=<PATH_TO_K3S>/contrib/gotests_templates` as an item.
```
Comment (briandowns) [inline tests/TESTING.md:24]:
```suggestion
and add `--template_dir=<PATH_TO_K3S>/contrib/gotests_templates` as an item. `gotests` also has documented integrations for all of the popular editors used to write Go.
```

Comment (briandowns) [reply tests/TESTING.md:24]: ...or something to that effect.

File: tests/TESTING.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+```bash
+go test ./tests/... -run E2E
+```
+
+## Contributing New Tests
+
+___
+We gladly accept new tests of all types. If you wish to create
+a new test, please submit a PR with a title that includes the words `<NAME_OF_TEST> Created`. Additionally, the final merge commit into K3s should also include those keywords. This
```
Comment (briandowns) [inline tests/TESTING.md:103]: We may want to consider creating a new label for these issues/PR's to be easier to identify or report on. /cc @cwayne18

Comment (dereknola) [reply tests/TESTING.md:103]: That would be a good idea. These rules for title and commits come from Shylaja on how QA wants to process and track testing.

Comment (briandowns) [reply tests/TESTING.md:103]: We may also then want to include a link to documentation that outlines those guidelines.

File: pkg/etcd/etcd_int_test.go (hunk 1 above)
Comment (dereknola) [reply pkg/etcd/etcd_int_test.go]: Fixed, added alias to differentiate from other util packages.

File: tests/TESTING.md (hunk 5 above)
Comment (dereknola) [reply tests/TESTING.md]: Added.

File: tests/TESTING.md (hunk 6 above)
Comment (dereknola) [reply tests/TESTING.md:103]: Holding off on submitting PR until next design call on this issue.

**Review by briandowns — APPROVED** [review]

**Comment by dereknola** [comment]
Closing this PR, it will be spit into two different PRs, one with just docs changes, the other with the testing cleanup.

**Pull request closed without merging**